<h3 style="color: #001a79;">Exercise 1.1</h3>

<hr style="border-top: 1px solid #001a79;" />

The above gives about a 1.5% chance of randomly selecting the correct cups. Calculate the minimum number of cups of tea required to ensure the probability of randomly selecting the correct cups is less than or equal to 1%.



In [1]:
# The following was adapted from here: https://stackoverflow.com/q/4941753

In [2]:
# Math Module Import
import math

From this address, https://www.w3schools.com/python/module_math.asp, math.comb() - Returns the number of ways to choose k items from n items without repetition and order. Imagine the progress Fischer would have made in his chosen career if computing and clever packages such as python / math was made available to him. Kind of blows the mind. 

In order to utilize this math function - to deciher number of combinations within the number of items, 2 arguments need to be passed to it. The first is the number of items to choose from, in our initial case below, it is 8 - because Fishers Lady Tea Experiment and our own analysis requires these specifics. 

The 2nd argument is the number of possibilities to choose - in our initial case it is 4 - simply because Fischers Lady Tea Experiment was based on the lady choosing correctly 4 cups made with tea in first or milk in first. 

In [3]:
# Number of ways of selecting 4 cups from 8
math.comb(8,4)

70

The result above is 70 - meaning there are 70 different combinations of choosing 4 items from 8 items. If we were to use binary numbering scheme to demonstrate simply with 3 items, then there would be 8 combinations. However to only be able to select 2 cups from the 3 presented leaves a far smaller number:

Combination

0  0  0

0  0  1

0  1  0

0  1  1

1  0  0

1  0  1

1  1  0

1  1  1


Lines containing 011, 101 and 110 are showing the possible combination if there were 3 cups of tea in the experiment and the lady had to guess which 2 had the tea in first out of the possible 3. 

In [4]:
#Probability, of randomly selecting correctly 2 cups out of 3:
math.comb(3, 2)

3

In [5]:
#Probability, as a  percentage, of randomly selecting the correct 2 cups out of a possible 3 cups to 2 decimal places is:
round((1.0 / math.comb(3, 2) * 100), 2)

33.33

Moving into slightly larger numbers, on this occasion we are checking the probability of the lady selecting the correct cups of tea from 4 possibilities, with 2 milk in first and 2 tea in first: 

In [6]:
# Probability of corectly selecting 2 from 4 cups of tea
math.comb(4,2)

6

In [7]:
#Probability, as a  percentage, of randomly selecting correct cups to 2 decimal places is:
round((1.0 / math.comb(4, 2) * 100), 2)

16.67

In [8]:
#Probability, as a  percentage, of randomly selecting correct cups to 2 decimal places is:
round((1.0 / math.comb(8, 4) * 100), 2)

1.43

In [9]:
# NUmber of ways of selecting 5 cups from 10.
math.comb(10,5)

252

In [10]:
# Probability, as a  percentage, of randomly selecting correct cups to 2 decimal places is:
round((1.0 / math.comb(10, 5) * 100), 2)

0.4

So,  if the experiment was performed with 10 cups total, five with milk in first and five with tea in first, then the chances of selecting correctly randomly is approximately 0.4%.


BACK TO THE ORIGINAL QUESTION: Calculate the minimum number of cups of tea required to ensure the probability of randomly selecting the correct cups is less than or equal to 1%

In [20]:
# Probability, as a  percentage, of randomly selecting from 8 cups correct cups to 2 decimal places is:
round((1.0 / math.comb(8, 4) * 100), 2)

1.43

In [19]:
# Probability, as a  percentage, of randomly selecting from 9 cups correct cups to 2 decimal places is:
round((1.0 / math.comb(9, 4) * 100), 2)

0.79

In [18]:
# Probability, as a  percentage, of randomly selecting from 10 cups correct cups to 2 decimal places is:
round((1.0 / math.comb(10, 4) * 100), 2)

0.48

With 8 cups and we get a percentage of 1.43%
With 9 cups and we get a percentage of 0.79%
With 10 cups and we get a percentage of 0.48%
Therefore the answer to 1.1 is : The minimum number of cups for the probability of randomly selecting the correct cups to be equal or less than 1% is equal to 9.


*Bonus:* How many would be required if you were to let the taster get one cup wrong while maintaining the 1% threshold?

To allow the taster to get one cup wrong while maintaining the threshold. SO basically the entered value into the math.comb() function call will need to be x and n-3 - where x is the unknown number of cups required and n being the number of correct number of selected cups, which we know to be 3 in this case. 
We can use trial end error to calculate our percentage probability as above. 
SO lets start with 8 cups and a rnadom selection of 3:

In [21]:
# Probability, as a  percentage, of randomly selecting 3 correct from 8 cups correct cups to 2 decimal places is:
round((1.0 / math.comb(8, 3) * 100), 2)

1.79

In [23]:
# Probability, as a  percentage, of randomly selecting 3 correct from 9 cups correct cups to 2 decimal places is:
round((1.0 / math.comb(9, 3) * 100), 2)

1.19

From 9 cups our selection has resulted in a  probability of 1.19%, so one more adjustment should get us below 1%. THis time using 10 cups:

In [24]:
# Probability, as a  percentage, of randomly selecting 3 correct from 10 cups correct cups to 2 decimal places is:
round((1.0 / math.comb(10, 3) * 100), 2)

0.83

Exercise 2
Remember to do these exercises in your own notebook in your assessment repository.

Use scipy's version of Fisher's exact test to simulate the Lady Tasting Tea problem.


In [25]:
# import SCIPY STAT - Statistical Function
import scipy.stats as ss

Table
             With Tea In | With Milk In
With Tea In      4               0
With Milk In     0               4


In [ ]:
Reference made to this address: https://www.statology.org/fishers-exact-test-python/

In [30]:
data = [[4,0],[0,4]]
print(ss.fisher_exact(data))

(inf, 0.028571428571428536)


In [ ]:
Additional reference to this address: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fisher_exact.html

In [26]:
ss.fisher_exact([[4, 0], [0, 4]])

(inf, 0.028571428571428536)